File: api-service/app/services/mimo_ota/quiet_zone_evidence.py
"""Fail-closed quiet-zone evidence contract for MIMO OTA consumers."""

from __future__ import annotations

import math
from typing import Any, Dict, Optional


QUIET_ZONE_EVIDENCE_SCHEMA_VERSION = 1

_KEYS = {
    "schema_version",
    "status",
    "source",
    "formal_verified",
    "measured_ripple_db",
    "proxy_ripple_db",
    "calibration_id",
}


def _finite_number(value: Any) -> Optional[float]:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return None
    number = float(value)
    return number if math.isfinite(number) else None


def build_quiet_zone_evidence(proxy_ripple_db: Any) -> Dict[str, Any]:
    """Build the only two states currently supported by an authoritative writer.

    ProbePattern peak spread is useful diagnostics, but it is not a multi-point
    quiet-zone field measurement.  The real grid acquisition path remains
    fail-closed until a sourced centimetre-unit linear-stage contract exists.
    """
    proxy = _finite_number(proxy_ripple_db)
    return {
        "schema_version": QUIET_ZONE_EVIDENCE_SCHEMA_VERSION,
        "status": "diagnostic_proxy" if proxy is not None else "unavailable",
        "source": "probe_pattern_peak_spread" if proxy is not None else "missing",
        "formal_verified": False,
        "measured_ripple_db": None,
        "proxy_ripple_db": proxy,
        "calibration_id": None,
    }


def parse_quiet_zone_evidence(value: Any) -> Optional[Dict[str, Any]]:
    """Return a normalized snapshot only for an exact writer-supported state."""
    if not isinstance(value, dict) or set(value) != _KEYS:
        return None
    if (
        type(value.get("schema_version")) is not int
        or value.get("schema_version") != QUIET_ZONE_EVIDENCE_SCHEMA_VERSION
    ):
        return None
    if value.get("formal_verified") is not False:
        return None
    if value.get("measured_ripple_db") is not None:
        return None
    if value.get("calibration_id") is not None:
        return None

    status = value.get("status")
    source = value.get("source")
    if status == "unavailable" and source == "missing":
        if value.get("proxy_ripple_db") is not None:
            return None
        return build_quiet_zone_evidence(None)
    if status == "diagnostic_proxy" and source == "probe_pattern_peak_spread":
        proxy = _finite_number(value.get("proxy_ripple_db"))
        if proxy is None:
            return None
        return build_quiet_zone_evidence(proxy)
    return None


def quiet_zone_evidence_is_formally_verified(value: Any) -> bool:
    """Current supported states are explicitly non-formal."""
    parsed = parse_quiet_zone_evidence(value)
    return parsed is not None and parsed["formal_verified"] is True


def quiet_zone_scope_is_formally_verified(precheck: Any) -> bool:
    """Legacy flags and source strings cannot promote an untrusted snapshot."""
    if not isinstance(precheck, dict):
        return False
    return quiet_zone_evidence_is_formally_verified(
        precheck.get("quiet_zone_evidence")
    )

File: gui/src/components/Commissioning/quietZoneEvidence.ts
type QuietZoneEvidence = {
  schema_version: 1
  status: 'unavailable' | 'diagnostic_proxy'
  source: 'missing' | 'probe_pattern_peak_spread'
  formal_verified: false
  measured_ripple_db: null
  proxy_ripple_db: number | null
  calibration_id: null
}

const EVIDENCE_KEYS = [
  'calibration_id',
  'formal_verified',
  'measured_ripple_db',
  'proxy_ripple_db',
  'schema_version',
  'source',
  'status',
].sort()

function parseQuietZoneEvidence(value: unknown): QuietZoneEvidence | null {
  if (!value || typeof value !== 'object' || Array.isArray(value)) return null
  const record = value as Record<string, unknown>
  if (Object.keys(record).sort().join('|') !== EVIDENCE_KEYS.join('|')) return null
  if (record.schema_version !== 1 || record.formal_verified !== false) return null
  if (record.measured_ripple_db !== null || record.calibration_id !== null) return null

  if (
    record.status === 'unavailable'
    && record.source === 'missing'
    && record.proxy_ripple_db === null
  ) return record as QuietZoneEvidence

  if (
    record.status === 'diagnostic_proxy'
    && record.source === 'probe_pattern_peak_spread'
    && typeof record.proxy_ripple_db === 'number'
    && Number.isFinite(record.proxy_ripple_db)
  ) return record as QuietZoneEvidence

  return null
}

export function describePrecheckOutcome(
  value: unknown,
  evidenceValue: unknown,
  operationalReady: unknown,
) {
  const formalVerified = parseQuietZoneEvidence(evidenceValue)?.formal_verified as boolean | undefined
  if (value === true && formalVerified === true) {
    return {
      color: 'green',
      title: '预检通过',
      message: '所有系统、校准与静区测量证据均满足正式测试要求。',
    }
  }
  if (value === false && operationalReady === false) {
    return {
      color: 'red',
      title: '预检失败',
      message: '系统存在异常，请检查以下信息。',
    }
  }
  return {
    color: 'yellow',
    title: '预检未判定',
    message: '运行条件已检查，但缺少正式静区测量证据。诊断流程可继续。',
  }
}

export function describePrecheckMessages(value: unknown, evidenceValue: unknown): string[] {
  const formalVerified = parseQuietZoneEvidence(evidenceValue)?.formal_verified as boolean | undefined
  if (formalVerified === true && Array.isArray(value)) {
    return value.filter((message): message is string => typeof message === 'string')
  }
  return [
    '静区结论未判定：无权威多点场扫描证据；历史提示未作为正式证据发布。',
  ]
}

export function describeQuietZoneEvidence(value: unknown) {
  const evidence = parseQuietZoneEvidence(value)
  const proxy = evidence?.status === 'diagnostic_proxy'
    ? evidence.proxy_ripple_db
    : null
  return {
    verified: false,
    formalRipple: 'N/A',
    proxyRipple: typeof proxy === 'number' ? `${proxy.toFixed(2)} dB` : null,
    label: proxy !== null
      ? 'ProbePattern 峰值离散诊断代理，非静区实测'
      : '无权威多点场扫描证据',
  }
}
